File: training-data/go/jump_game.go
package main

import "fmt"

func canJump(a []int) bool {
	reach := 0
	for i := 0; i < len(a) && i <= reach; i++ {
		reach = max(reach, i+a[i])
	}
	return reach >= len(a)-1
}

func minJumps(a []int) int {
	jumps, end, far := 0, 0, 0
	for i := 0; i < len(a)-1; i++ {
		far = max(far, i+a[i])
		if i == end {
			if far <= i {
				return -1
			}
			jumps++
			end = far
		}
	}
	return jumps
}

func main() {
	fmt.Println(canJump([]int{2, 3, 1, 1, 4}), canJump([]int{3, 2, 1, 0, 4}))
	fmt.Println(minJumps([]int{2, 3, 1, 1, 4}), minJumps([]int{3, 2, 1, 0, 4}))
}

File: training-data/rust/jump_game.rs
fn can_jump(a: &[usize]) -> bool {
    let mut reach = 0;
    for (i, &x) in a.iter().enumerate() {
        if i > reach {
            return false;
        }
        reach = reach.max(i + x);
    }
    true
}

fn min_jumps(a: &[usize]) -> Option<usize> {
    let (mut jumps, mut end, mut far) = (0, 0, 0);
    for i in 0..a.len().saturating_sub(1) {
        far = far.max(i + a[i]);
        if i == end {
            if far <= i {
                return None;
            }
            jumps += 1;
            end = far;
        }
    }
    Some(jumps)
}

fn main() {
    println!("{} {}", can_jump(&[2, 3, 1, 1, 4]), can_jump(&[3, 2, 1, 0, 4]));
    println!("{:?} {:?}", min_jumps(&[2, 3, 1, 1, 4]), min_jumps(&[3, 2, 1, 0, 4]));
}
